brave/brave-core
# Implement Verifiable Advertiser Conversions for Brave Ads

Resolves https://github.com/brave/brave-browser/issues/13368

## Submitter Checklist:


## Reviewer Checklist:


## After-merge Checklist:

  changes has landed on.

## Test Plan:

0. start browser on fresh profile with `npm start -- --enable-logging=stderr --vmodule="*/bat-native-ads/*"=6,"*/brave_ads/*"=6,"*/brave_user_model/*"=6,"*/bat_ads/*"=6 --brave-ads-staging --rewards=staging=true --brave-ads-debug`

1. run test site with conversion tag (e.g. `<meta name="ad-conversion-id" content="smartbrownfoxes42">`) on `localhost:8000`

2. add test "untargeted" ad with conversion ID and test key to catalog and inject via proxy, e.g.

```
 "conversions": [
            {
              "type": "postview",
              "urlPattern": "http://localhost:8000/*",
              "observationWindow": 1,
              "conversionPublicKey": "ofIveUY/bM7qlL9eIkAv/xbjDItFs1xRTTYKRZZsPHI="
            }
          ]
```

3. trigger add of segment "untargeted"

4. navigate to `localhost:8000` and verify that a conversion was triggered with logs

```
[86997:775:0216/065612.681696:VERBOSE1:conversions.cc(412)] Convert creative instance id 4983dc92-09fd-4e5c-94c3-616a4728c1d9 and creative set id 5bdeab83-048f-48a7-9602-a1092ded123c in 0 hours, 3 minutes, 57 seconds at 07:00
[86997:775:0216/065612.684147:VERBOSE6:conversions.cc(306)] Successfully logged conversion event
```

5. wait until the confirmation for the conversion event is triggered (time in log above) and verify "envelope" payload via proxy

```
"envelope": {
"alg": "alg_placeholder",
"ciphertext": "ciphertext_placeholder",
"epk": "epk_placeholder",
"nonce": "nonce_placeholder"
},
```

6. Verify that payload can be decrypted via https://github.com/brave-experiments/VAC-Encryption-Envelopes/blob/main/src/advertiser.ts
